### PR TITLE
Snafu errors

### PR DESCRIPTION
**Pull request by HenningTimm**
The goal of this PR is to write more specific error messages using the snafu crate. This is a lot of work and I would appreciate help from anyone that would like to gain experience with the snafu crate.

## Why would I do that?
Until now, errors were unspecific. For example in `call_consensus_reads`, the error message only contained the info that a file couldn't be opened. Not which one of the 4-5 files was the culprit.

## What has already been done?

I started by adding an error module and slowly replacing the Result types in the `call_consensus_reads` module with custom error types. The variants of the `errors::Error` enum should each model a specific error, for example 
```
pub enum Error {
    #[snafu(display("Error parsing the following Starcode cluster in csv format {:?}: {}", record, source))]
    StarcodeClusterParseError {
        record: csv::StringRecord,
        source: csv::Error,
    },
    // ...
}
```
models the CSV reader failing to parse a cluster returned by starcode.

## What do I need to change

1. Add error types that model type and context of specific errors in your code to the `Error` enum. If reasonable, reuse an existing error type (for example for `FastqReaderError`) or merge existing quickerror types into there.
2. Wrap errors of libraries using the `context` function: For example:
    ```
    let fastq_reader = fastq::Reader::from_file(fq1).context(errors::FastqReaderError {
                            filename: String::from(fq1),
                        })?
    ```
    Basically, squeeze a `.context(<constructur for specific error type>)` in front of existing `?` operators.
    Note that this is not required for `?` calls that already return the `errors::Error` type.
3. If all errors in a function or method have been wrapped, replace the return type with `error::Result`.

### BAM 
- [x] depth.rs
- [x] call_consensus_reads.rs
    - [x] mod.rs
    - [x] pipeline.rs
    - [x] calc_consensus.rs
### BCF
- [x] baf.rs
- [x] match_variants.rs
- [x] to_txt.rs
### FASTQ
- [x] filter.rs
- [x] split.rs
- [x] call_consensus_reads.rs
    - [x] mod.rs
    - [x] pipeline.rs
    - [x] calc_consensus.rs
### Other
- [x] main.rs

### CODE REVIEW

**Comment by FelixMoelder** [comment]
To propagate an error in `match_variants.rs` to a higher level it was necessary to resolve a closure.
While the `matches`-function previously returned a `Option<u32>` it needed to be wrapped into a `Result<Option<u32>>`.

This is not only a bit shabby but also required additional changes as it does not seem to be possible to push the error to a higher level when the `matches`-function is called.

The code [replacement](https://github.com/rust-bio/rust-bio-tools/blob/a167683b5b66c0e1915711c83950c2d7c3a60d55/src/bcf/match_variants.rs#L117-L133) appears awful to me, but I could not find a better solution.
The code of the original closure can be found [here](https://github.com/rust-bio/rust-bio-tools/blob/8266a88c43a125ae572f19353d04456c986e8d2d/src/bcf/match_variants.rs#L95-L109).

File: src/bam/call_consensus_reads/pipeline.rs [hunk 1]
```diff
@@ -26,13 +27,13 @@ type RecordIDS = Vec<RecordID>;
...
 impl CallConsensusRead {
-    pub fn call_consensus_reads(&mut self) -> Result<(), Box<dyn Error>> {
+    pub fn call_consensus_reads(&mut self) -> errors::Result<()> {
         let mut group_end_idx: BTreeMap<Position, GroupIDs> = BTreeMap::new();
         let mut duplicate_groups: HashMap<GroupID, RecordIDS> = HashMap::new();
         let mut record_storage: HashMap<RecordID, RecordStorage> = HashMap::new();
 
         for (i, result) in self.bam_reader.records().into_iter().enumerate() {
-            let mut record = result?;
+            let mut record = result.context(errors::BamReadError { record: None })?;
```
Comment (johanneskoester) [inline src/bam/call_consensus_reads/pipeline.rs]: Why is there at all a record in BamReadError?

Comment (johanneskoester) [reply src/bam/call_consensus_reads/pipeline.rs]: Maybe rather add `i` to the error, so that people can see which record caused the error?

File: src/bam/call_consensus_reads/pipeline.rs [hunk 2]
```diff
@@ -65,7 +70,13 @@ impl CallConsensusRead {
                         Some(record_pair) => {
                             //For right record save end position and duplicate group ID
                             group_end_idx
-                                .entry(record.cigar_cached().unwrap().end_pos()? - 1)
+                                .entry(
+                                    record.cigar_cached().unwrap().end_pos().context(
+                                        errors::BamCigarError {
+                                            record: record.clone(),
```
Comment (johanneskoester) [inline src/bam/call_consensus_reads/pipeline.rs]: This clone will happen always, causing unnecessary work if no error occurs. Can context take a closure or something like that?

File: src/bcf/match_variants.rs [hunk 3]
```diff
@@ -99,19 +118,27 @@ pub fn match_variants(
...
 
-            rec.push_info_integer(b"MATCHING", &matching)?;
-            outbcf.write(&rec)?;
+            rec.push_info_integer(b"MATCHING", &matching)
+                .context(errors::BCFTagWriteError {
+                    data: format!("{:?}", matching),
+                    fd: String::from("MATCHING"),
+                })?;
+            outbcf.write(&rec).context(errors::BCFWriteError {
+                record: format!("{:?}", rec),
```
Comment (johanneskoester) [inline src/bcf/match_variants.rs]: Can the formatting be automated inside the error? Why is it needed here, and not above?

File: src/bam/call_consensus_reads/pipeline.rs (hunk 1 above)
Comment (FelixMoelder) [reply src/bam/call_consensus_reads/pipeline.rs]: You are right. The record does not make sense at this place. I'll replace it by its index.

File: src/bam/call_consensus_reads/pipeline.rs (hunk 2 above)
Comment (FelixMoelder) [reply src/bam/call_consensus_reads/pipeline.rs]: That has now been changed to return the cigar string itself instead the record.

File: src/bcf/match_variants.rs (hunk 3 above)
Comment (FelixMoelder) [reply src/bcf/match_variants.rs]: As discussed records have been removed from write errors output messages.

**Comment by HenningTimm** [comment]
When starting this, I decided to put all errors into one `Enum` so I could reuse them in different modules. This resulted in a huge and unwieldy Error enum and, as @johanneskoester suggested, makes it harder to practice good error handling. Additionally, the [philosophy of snafu](https://docs.rs/snafu/0.4.1/snafu/guide/philosophy/index.html#many-error-types) also suggests that splitting up errors is nice. So, before merging this, we should think about splitting the `errors.rs` up and put one into each module.

This would require us to handle all errors in the main function. However, since all errors we use have a reasonable error message, we could print that and return some generic error in main.

**Comment by johanneskoester** [comment]
Or just return `Box<Error>` in main.

**Comment by johanneskoester** [comment]
I have just released new versions of rust-bio and rust-htslib with snafu errors. In the end, I decided to have one error enum per top-level module. It seemed like a good compromise for an API. However, here, I think one big error enum is fine, because they are anyway only presented to the user. There is no need to handle them on an API level. Hence, @HenningTimm may I ask you to just adapt to the latest rust-bio and rust-htslib release? Then, I'll be able to merge this.

**Comment by HenningTimm** [comment]
@johanneskoester I adapted the snafu errors to work with the new underlying error types. I had to change the behaviour of `main.rs` slightly to actually display them. 

Just returning the error (using a main function with ` -> errors::Result<()> {`) prints the debug version of the error, i.e. the struct with its fields, not the rendered error as specified with the `#[snafu(display()...)]` decorator. 

Currently the error is first printed as a formatted string and then returned by `main` and printed again in debug form. While this provides two (potentially different) views on the error, I am not happy with this. Dropping the rendered string makes the error hard to read, dropping the return obfuscates error states/ return codes.

**Comment by HenningTimm** [comment]
While I was already touching the code, I added some other clippy lints.

**Comment by johanneskoester** [comment]
Closing in favor of the new anyhow and thiserror implementation.

**Comment by johanneskoester** [comment]
Nevertheless, thanks a lot for this valuable prototyping work!
